pkg/controller/persistentvolume/framework_test.go

```diff
@@ -471,6 +471,62 @@ func (r *volumeReactor) waitTest() {
 	}
 }
 
+// deleteVolumeWithEvent simulates that a volume has been deleted in etcd and
+// the controller receives 'volume deleted' event.
+func (r *volumeReactor) deleteVolumeWithEvent(volume *api.PersistentVolume) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// Remove the volume from list of resulting volumes.
+	delete(r.volumes, volume.Name)
+
+	// Generate deletion event. Cloned volume is needed to prevent races (and we
+	// would get a clone from etcd too).
+	clone, _ := conversion.NewCloner().DeepCopy(volume)
+	volumeClone := clone.(*api.PersistentVolume)
+	r.volumeSource.Delete(volumeClone)
+}
+
+// deleteClaimWithEvent simulates that a claim has been deleted in etcd and the
+// controller receives 'claim deleted' event.
+func (r *volumeReactor) deleteClaimWithEvent(claim *api.PersistentVolumeClaim) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// Remove the claim from list of resulting claims.
+	delete(r.claims, claim.Name)
+
+	// Generate deletion event. Cloned volume is needed to prevent races (and we
+	// would get a clone from etcd too).
+	clone, _ := conversion.NewCloner().DeepCopy(claim)
```

Comment:
I am wondering whether the clone operation in this func and deleteVolumeWithEvent function can be put before the lock?